### PR TITLE
Implement HedgingResilienceStrategy

### LINKED ISSUE #876: Introduce Hedging Strategy
We are working on the development of a generic `HedgingPolicy` to use as part of our `HttpClient` Resilience pipeline.
Since the hedging as a policy is not yet supported by Polly we consider setting up a new package dedicated to the generic implementation of such policy. Please let us know if you are interested in having this and how we could proceed further.

*Summary*
The hedging policy will allow the sequential and delayed execution of a set of tasks, until one of them completes successfully or until all of them are completed. 
First task that is successfully completed triggers the cancellation and disposal of all other tasks and/or adjacent allocated resources. 

General structure of polly policy:
- The `Async<PolicyName>Syntax` exposes extension method for the `PolicyBuilder` which builds new instances of the `Async<PolicyName>`:
   - e.g. for Hedging: `AsyncHedgingSyntax` exposes methods to create `AsyncHedgingPolicy` instances.
- The generic `Async<PolicyName>Policy` is the actual policy, which:
  - Inherits the `AsyncPolicy<TResult>` base class
  - Exposes internal constructor and it is created only via the  `Async<PolicyName>Syntax` extensions.
  - Implements the abstract method of the base class, ` Async<PolicyName>Syntax` which contains the actual implementation, i.e.
    ```
    Task<TResult> ImplementationAsync(
        Func<Context, CancellationToken, Task<TResult>> action,
        Context context, 
        CancellationToken cancellationToken,
        bool continueOnCapturedContext)
    ```
    - e.g. for Hedging: `AsyncHedgingPolicy`
________________________________________________________________________________

## Hedging policy configuration
The base implementation will take:
- `HedgingDelay` - how long shall one wait before spawning a new hedged request;
- `OnHedgingAsync` - task performed after every spawning a hedged request; [can be used for logging & telemetry]
- A list of `hedgedTaskFunctions` - the tasks performed via hedging if the primary execution delegate fails.

____________________________________________________________________________
## Hedging logic
Steps:
1. Start executing task [first one first] 
2. Wait up to HedgingDelay ms 
3. If the request is: 
  - A. Completed: 
    - If result is an exception 
      - If ShouldHandleException - GOTO [B] 
      - else - GOTO [4.] 
    - If result is not exception 
      - If ShouldHandleResult - GOTO [B] 
      - else - GOTO [4.] (e.g. result is successful, so return result).
  - B. Not Completed: 
    - Next task will be executed - GOTO [1.] 

4. The result of the execution can be: 
  - [Success]
    - First request to complete successfully triggers the cancellation of all others
    - Return successful result.
  - [Failure]
    - No request completed successfully, and all requests have been executed. 
    - Return the failure of the first task. 
____________________________________________________________________________

## Exposed API for policy creation:
```
 var hedgingPolicy = Policy
                .Handle<Exception>()
                .OrResult<string>(_ => false)
                .AsyncHedgingPolicy(
                    hedgedTaskFunctions,
                    defaultHedgingDelay,
                    onHedgingAsync);

```

### PR DESCRIPTION
**Pull request by martintmk**
### The issue or feature being addressed

Closes #876

### Details on the issue fix or feature implementation

This PR adds all necessary infrastructure and implements Hedging Resilience Strategy. 

- When the callbacks are hedged each execution gets its own `ResilienceContext` and `CancellationToken` so the concurrent calls won't interfere with each other and can be cancelled individually. 
- Once any callback gets acceptable result the properties from cloned `ResilienceContext` are merged bac to the main one.
- The primary callback still re-uses the main context.

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Comment by martintmk** [comment]
I'll add more tests , but the implementation is finished from my perspective.

File: src/Polly.Core.Tests/Hedging/Controller/HedgingExecutionContextTests.cs [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+using System.Globalization;
+using System.Threading.Tasks;
+using Polly.Core.Tests.Helpers;
+using Polly.Hedging;
+using Polly.Hedging.Controller;
+using Polly.Hedging.Utils;
+using Polly.Strategy;
+using Polly.Utils;
+
+namespace Polly.Core.Tests.Hedging.Controller;
```
Comment (martincostello) [inline src/Polly.Core.Tests/Hedging/Controller/HedgingExecutionContextTests.cs:11]:
```suggestion
namespace Polly.Core.Tests.Hedging.Controller;

```

File: src/Polly.Core.Tests/Hedging/Controller/TaskExecutionTests.cs [hunk 2]
```diff
@@ -0,0 +1,269 @@
+using System;
+using System.Threading.Tasks;
+using Polly.Core.Tests.Helpers;
+using Polly.Hedging;
+using Polly.Hedging.Controller;
+using Polly.Hedging.Utils;
+using Polly.Strategy;
+
+namespace Polly.Core.Tests.Hedging.Controller;
```
Comment (martincostello) [inline src/Polly.Core.Tests/Hedging/Controller/TaskExecutionTests.cs:9]:
```suggestion
namespace Polly.Core.Tests.Hedging.Controller;

```

File: src/Polly.Core/Hedging/Controller/HedgedTaskType.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Polly.Hedging.Utils;
+#pragma warning disable CA1031 // Do not catch general exception types
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/HedgedTaskType.cs]:
```suggestion
```

File: src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+    private readonly List<TaskExecution> _tasks = new();
+    private readonly List<TaskExecution> _executingTasks = new();
+    private readonly IObjectPool<TaskExecution> _executionPool;
+    private readonly TimeProvider _timeProvider;
+    private readonly int _maxAttempts;
+    private readonly Action<HedgingExecutionContext> _onReset;
+    private readonly ResilienceProperties _replacedProperties = new();
+
+    public HedgingExecutionContext(IObjectPool<TaskExecution> executionPool, TimeProvider timeProvider, int maxAttempts, Action<HedgingExecutionContext> onReset)
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs:29]:
```suggestion
    public HedgingExecutionContext(
        IObjectPool<TaskExecution> executionPool,
        TimeProvider timeProvider,
        int maxAttempts,
        Action<HedgingExecutionContext> onReset)
```

File: src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    public ContextSnapshot Snapshot { get; private set; }
+
+    public bool IsInitialized => Snapshot.Context != null;
+
+    public IReadOnlyList<TaskExecution> Tasks => _tasks;
+
+    private bool ContinueOnCapturedContext => Snapshot.Context.ContinueOnCapturedContext;
+
+    public async ValueTask<ExecutionInfo<TResult>> LoadExecutionAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> primaryCallback, TState state)
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs:54]:
```suggestion
    public async ValueTask<ExecutionInfo<TResult>> LoadExecutionAsync<TResult, TState>(
        Func<ResilienceContext,
        TState,
        ValueTask<TResult>> primaryCallback,
        TState state)
```

File: src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            {
+                index = i;
+                break;
+            }
+        }
+
+        if (index is not null)
+        {
+            var execution = _executingTasks[index.Value];
+            _executingTasks.RemoveAt(index.Value);
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs:200]:
```suggestion
            var indexValue = index.GetValueOrDefault();
            var execution = _executingTasks[indexValue];
            _executingTasks.RemoveAt(indexValue);
```

File: src/Polly.Core/Hedging/Controller/TaskExecution.cs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    public void Cancel()
+    {
+        if (IsAccepted)
+        {
+            return;
+        }
+
+        _cancellationSource!.Cancel();
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/TaskExecution.cs:60]:
```suggestion
        if (!IsAccepted)
        {
            _cancellationSource!.Cancel();
        }
```

File: src/Polly.Core/Hedging/HedgingResilienceStrategy.cs [hunk 8]
```diff
@@ -20,9 +37,74 @@ public HedgingResilienceStrategy(HedgingStrategyOptions options)
 
     public HedgingHandler.Handler? HedgingHandler { get; }
 
-    protected internal override ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> callback, ResilienceContext context, TState state)
+    public OutcomeEvent<OnHedgingArguments>.Handler? OnHedgingHandler { get; }
+
+    protected internal override async ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> callback, ResilienceContext context, TState state)
```
Comment (martincostello) [inline src/Polly.Core/Hedging/HedgingResilienceStrategy.cs:42]:
```suggestion
    protected internal override async ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(
        Func<ResilienceContext,
        TState, ValueTask<TResult>> callback,
        ResilienceContext context,
        TState state)
```

File: src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs (hunk 6 above)
Comment (martintmk) [reply src/Polly.Core/Hedging/Controller/HedgingExecutionContext.cs]: Simplified this piece.

File: src/Polly.Core/Hedging/HedgingResilienceStrategy.cs [hunk 9]
```diff
@@ -20,9 +37,77 @@ public HedgingResilienceStrategy(HedgingStrategyOptions options)
...
+
+        var continueOnCapturedContext = context.ContinueOnCapturedContext;
+        context.CancellationToken.ThrowIfCancellationRequested();
+
+        // create hedging execution context
+        var hedgingContext = _controller.GetContext(context);
+
+        try
+        {
+            while (true)
```
[the remark this answers is not in the document]
Comment (martintmk) [reply src/Polly.Core/Hedging/HedgingResilienceStrategy.cs:61]: kept while but put the cancellation token check immediately after (compiler complain about missing return value)

File: src/Polly.Core/Hedging/Controller/TaskExecution.cs [hunk 10]
```diff
@@ -9,6 +9,8 @@ namespace Polly.Hedging.Controller;
 /// <summary>
 /// Represents a single hedging attempt execution alongside all the necessary resources. These are:
 ///
+///  <ul>
+///  </ul>
 /// - Distinct <see cref="ResilienceContext"/> instance for this execution. One exception are primary task where the main context is reused.
 /// - The cancellation token associated with the execution.
```
Comment (martincostello) [inline src/Polly.Core/Hedging/Controller/TaskExecution.cs]:
```suggestion
///  <ul>
/// <li>Distinct <see cref="ResilienceContext"/> instance for this execution. One exception are primary task where the main context is reused.</li>
/// <li>The cancellation token associated with the execution.</li>
/// </ul>
```

File: src/Polly.Core/Hedging/Controller/TaskExecution.cs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+namespace Polly.Hedging.Controller;
+
+#pragma warning disable CA1031 // Do not catch general exception types
+
+/// <summary>
+/// Represents a single hedging attempt execution alongside all the necessary resources. These are:
+///
+/// - Distinct <see cref="ResilienceContext"/> instance for this execution. One exception are primary task where the main context is reused.
+/// - The cancellation token associated with the execution.
```
[the remark this answers is not in the document]
Comment (martintmk) [reply src/Polly.Core/Hedging/Controller/TaskExecution.cs]: i used the `<list>` + `<item>` here.

**Pull request merged**